keras_nlp/models/deberta/disentangled_self_attention.py

```diff
@@ -0,0 +1,376 @@
+        self.bucket_size = bucket_size
+        self.dropout = dropout
+
+        # Initializers.
+        self._kernel_initializer = keras.initializers.get(kernel_initializer)
+        self._bias_initializer = keras.initializers.get(bias_initializer)
+
+        # Derived args.
+        self.attn_head_size = hidden_dim // num_heads
+
+        num_type_attn = 3
+        self.scale_factor = 1.0 / math.sqrt(
+            float(num_type_attn * self.attn_head_size)
+        )
+
+        # Layers.
+
+        # Q, K, V linear layers.
+        self._query_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_heads, self.attn_head_size),
+            bias_axes="de",
+            **self._get_common_kwargs_for_sublayer(use_bias=True),
+            name="query",
+        )
+        self._key_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_heads, self.attn_head_size),
+            bias_axes="de",
+            **self._get_common_kwargs_for_sublayer(use_bias=True),
+            name="key",
+        )
+        self._value_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_heads, self.attn_head_size),
+            bias_axes="de",
+            **self._get_common_kwargs_for_sublayer(use_bias=True),
+            name="value",
+        )
+
+        # Relative attention.
+        self._position_dropout_layer = keras.layers.Dropout(self.dropout)
+
+        self._attn_dropout_layer = keras.layers.Dropout(
+            self.dropout, name="attention_dropout"
+        )
+        self._softmax = keras.layers.Softmax(axis=-1, name="attention_softmax")
+
+        # Output.
+        self._output_dense = keras.layers.EinsumDense(
+            equation="abc,cd->abd",
+            output_shape=(None, self.hidden_dim),
+            bias_axes="d",
+            **self._get_common_kwargs_for_sublayer(use_bias=True),
+            name="attention_output",
+        )
+
+    def _get_common_kwargs_for_sublayer(self, use_bias=True):
+        common_kwargs = {}
+
+        kernel_initializer = clone_initializer(self._kernel_initializer)
+        bias_initializer = clone_initializer(self._bias_initializer)
+
+        common_kwargs["kernel_initializer"] = kernel_initializer
+        if use_bias:
+            common_kwargs["bias_initializer"] = bias_initializer
+
+        return common_kwargs
+
+    def _masked_softmax(self, attention_scores, attention_mask=None):
+        """Normalizes the attention scores to probabilities using softmax."""
+
+        if attention_mask is not None:
+            # The expand dim happens starting from the `num_heads` dimension,
+            # (<batch_dims>, num_heads, <query_attention_dims,
+            # key_attention_dims>)
+            mask_expansion_axis = -3
+            for _ in range(
+                len(attention_scores.shape) - len(attention_mask.shape)
```

Comment:
does `attention_scores.shape.rank - attention_mask.shape.rank` work here? Would be more readable.